File: cloud-admin-portal/backend/app/services/reconciler.py
import asyncio
import logging
from datetime import datetime, timezone, timedelta

from sqlalchemy.orm import Session

from ..models.db import TenantDB
from .pd_client import PDClient
from .tenant import is_tenant_keyspace, parse_keyspace

log = logging.getLogger(__name__)


class Reconciler:
    def __init__(self, pd: PDClient, db_session_factory, interval_seconds: int = 300):
        self._pd = pd
        self._db_factory = db_session_factory
        self._interval = interval_seconds
        self._running = False
        self._cursor = ""

    async def start(self):
        self._running = True
        log.info("Reconciler started (interval=%ds)", self._interval)

        await asyncio.sleep(10)

        while self._running:
            try:
                self._tick()
            except Exception as e:
                log.error("Reconciler tick failed: %s", e)

            await asyncio.sleep(self._interval)

    def stop(self):
        self._running = False
        log.info("Reconciler stopped")

    def _tick(self):
        db: Session = self._db_factory()
        try:
            self._recover_stuck_tenants(db)
            self._incremental_sweep(db)
            db.commit()
        except Exception:
            db.rollback()
            raise
        finally:
            db.close()

    def _recover_stuck_tenants(self, db: Session):
        cutoff = datetime.now(timezone.utc) - timedelta(minutes=10)
        stuck = db.query(TenantDB).filter(
            TenantDB.state.in_(("CREATING", "DISABLING")),
            TenantDB.created_at < cutoff,
        ).all()

        for tenant in stuck:
            ks = self._pd.get_keyspace(tenant.keyspace)
            if tenant.state == "CREATING":
                if ks:
                    tenant.state = "ACTIVE"
                    tenant.state_reason = "Recovered by reconciler: keyspace exists"
                    log.info("Recovered stuck CREATING tenant %s -> ACTIVE", tenant.id)
                else:
                    tenant.state = "CREATE_FAILED"
                    tenant.state_reason = "Recovered by reconciler: keyspace not found after timeout"
                    log.info("Recovered stuck CREATING tenant %s -> CREATE_FAILED", tenant.id)
            elif tenant.state == "DISABLING":
                if ks is None or (isinstance(ks, dict) and ks.get("state") == "DISABLED"):
                    tenant.state = "DISABLED"
                    tenant.state_reason = "Recovered by reconciler"
                    log.info("Recovered stuck DISABLING tenant %s -> DISABLED", tenant.id)
                else:
                    self._pd.disable_keyspace(tenant.keyspace)
                    tenant.state = "DISABLED"
                    tenant.state_reason = "Recovered by reconciler: force-disabled"
                    log.info("Force-disabled stuck DISABLING tenant %s", tenant.id)

            tenant.updated_at = datetime.now(timezone.utc)

    def _incremental_sweep(self, db: Session):
        batch = (
            db.query(TenantDB)
            .filter(TenantDB.state == "ACTIVE", TenantDB.id > self._cursor)
            .order_by(TenantDB.id)
            .limit(100)
            .all()
        )

        if not batch:
            self._cursor = ""
            return

        for tenant in batch:
            ks = self._pd.get_keyspace(tenant.keyspace)
            if ks is None:
                log.warning(
                    "Tenant %s (keyspace=%s) is ACTIVE in DB but missing from PD",
                    tenant.id, tenant.keyspace,
                )

        self._cursor = batch[-1].id

    def sync_new_keyspaces(self, db: Session):
        tikv_keyspaces = self._pd.list_keyspaces()
        tikv_tenant_keyspaces = {}

        for ks in tikv_keyspaces:
            name = ks.get("name", "") if isinstance(ks, dict) else str(ks)
            if name and is_tenant_keyspace(name):
                parsed = parse_keyspace(name)
                if parsed:
                    tenant_id, _ = parsed
                    tikv_tenant_keyspaces[name] = tenant_id

        db_keyspaces = {t.keyspace for t in db.query(TenantDB).all()}
        new_keyspaces = set(tikv_tenant_keyspaces.keys()) - db_keyspaces
        created_count = 0

        for keyspace in new_keyspaces:
            tenant_id = tikv_tenant_keyspaces[keyspace]
            existing = db.query(TenantDB).filter(TenantDB.id == tenant_id).first()
            if existing:
                continue
            db.add(TenantDB(
                id=tenant_id,
                keyspace=keyspace,
                state="ACTIVE",
                created_at=datetime.now(timezone.utc),
                created_by="reconciler_sync",
            ))
            created_count += 1

        if created_count > 0:
            log.info("Reconciler synced %d new keyspaces from PD", created_count)

        return created_count

File: cloud-admin-portal/backend/src/services/reconciler.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

use sqlx::AnyPool;

use crate::db;
use crate::services::pd_client::PdClient;
use crate::{tenant_state, TENANT_ID_LEN};

pub struct Reconciler {
    pd: PdClient,
    db: AnyPool,
    interval: Duration,
    running: Arc<AtomicBool>,
}

impl Reconciler {
    pub fn new(pd: PdClient, db: AnyPool, interval_secs: u64) -> Self {
        Self {
            pd,
            db,
            interval: Duration::from_secs(interval_secs),
            running: Arc::new(AtomicBool::new(true)),
        }
    }

    pub fn stop_handle(&self) -> Arc<AtomicBool> {
        self.running.clone()
    }

    pub async fn sync_new_keyspaces(&self) -> usize {
        let keyspaces = self.pd.list_keyspaces().await;
        let existing = match db::get_all_keyspaces_from_db(&self.db).await {
            Ok(v) => v,
            Err(e) => { tracing::warn!("Failed to list DB keyspaces: {e}"); return 0; }
        };

        let existing_ks: std::collections::HashSet<String> = existing.iter().map(|(_, ks)| ks.clone()).collect();
        let mut count = 0;

        for ks_val in &keyspaces {
            let name = match ks_val.get("name").and_then(|v| v.as_str()) {
                Some(n) => n,
                None => continue,
            };
            if name == "DEFAULT" || name == "default" || existing_ks.contains(name) {
                continue;
            }
            let id = match name.strip_prefix(crate::KEYSPACE_PREFIX) {
                Some(stripped) if stripped.len() == TENANT_ID_LEN && stripped.chars().all(|c| c.is_ascii_alphanumeric()) => stripped,
                _ => continue,
            };
            let now = chrono::Utc::now().to_rfc3339();
            if db::insert_tenant(&self.db, id, name, tenant_state::ACTIVE, &now).await.is_ok() {
                tracing::info!("Synced keyspace from PD: {name} -> tenant {id}");
                count += 1;
            }
        }
        count
    }

    pub async fn start(self) {
        while self.running.load(Ordering::Relaxed) {
            tokio::time::sleep(self.interval).await;
            if !self.running.load(Ordering::Relaxed) {
                break;
            }
            self.run_cycle().await;
        }
    }

    async fn run_cycle(&self) {
        let cutoff = (chrono::Utc::now() - chrono::Duration::minutes(10)).to_rfc3339();

        if let Ok(stuck) = db::get_stuck_tenants(&self.db, tenant_state::CREATING, &cutoff).await {
            for t in &stuck {
                let exists = self.pd.get_keyspace(&t.keyspace).await.is_some();
                let (new_state, reason) = if exists {
                    (tenant_state::ACTIVE, "Recovered by reconciler: keyspace exists")
                } else {
                    (tenant_state::CREATE_FAILED, "Recovered by reconciler: keyspace not found")
                };
                if let Err(e) = db::update_tenant_state(&self.db, &t.id, new_state, Some(reason)).await {
                    tracing::warn!("Failed to recover tenant {}: {e}", t.id);
                }
            }
        }

        if let Ok(stuck) = db::get_stuck_tenants(&self.db, tenant_state::DISABLING, &cutoff).await {
            for t in &stuck {
                if let Err(e) = db::update_tenant_state(&self.db, &t.id, tenant_state::DISABLED, Some("Recovered by reconciler")).await {
                    tracing::warn!("Failed to recover tenant {}: {e}", t.id);
                }
            }
        }
    }
}
